source/citnames/source/Output.cc

```diff
@@ -81,49 +82,81 @@ namespace {
         const cs::Content config;
     };
 
-    // Duplicate detection filter.
-    //
-    // Duplicate detection can be done two ways: first one is based on the `output` attribute,
-    // second is based on all attributes. The benefit to use the `output` attribute only is,
-    // that it faster and handles compiler flag changes better. The problem with it is, that
-    // it not present in every entries.
-    //
-    // Current implementation is starts with the output one, but computes both hashes (and
-    // maintains the hash sets). If an entry has no output attribute, then it switch to use
-    // all attributes hashes.
+    using DuplicateFilterPtr = std::unique_ptr<struct DuplicateFilter>;
+
     struct DuplicateFilter : public Filter {
-        DuplicateFilter()
-                : hashes()
-        { }
+        static DuplicateFilterPtr from_content(const cs::Content&);
 
         bool apply(const cs::Entry &entry) override {
-            if (const auto h2 = hash(entry); hashes.find(h2) == hashes.end()) {
-                hashes.insert(h2);
-                return true;
-            }
-            return false;
+            const auto h2 = hash(entry);
+            auto [_, new_entry] = hashes.emplace(h2);
+            return new_entry;
         }
 
     private:
-        // The hash function based on all attributes.
-        //
-        // - It shall ignore the compiler name, but count all compiler flags in.
-        // - Same compiler call semantic is detected by filter out the irrelevant flags.
-        static std::string hash(const cs::Entry &entry) {
-            auto file = entry.file.string();
-            std::reverse(file.begin(), file.end());
-
-            const auto args = fmt::format(
-                    "{}",
-                    fmt::join(entry.arguments.rbegin(), std::prev(entry.arguments.rend()), ","));
-            size_t args_hash = std::hash<std::string>{}(args);
-
-            return fmt::format("{}:{}", args_hash, file);
-        }
+        virtual size_t hash(const cs::Entry&) = 0;
```

Comment:
Can this method be a const method? (I guess it won't change the filter object.)

Reply:
Yes you're right